pysnark/runtime.py

```diff
@@ -306,19 +306,46 @@ def _check_both_bits(self, other):
         raise ValueError("not a bit: " + str(other))
     
     def __and__(self, other):
-        """ Bitwise and &. Cost: 1 constraint """
-        self._check_both_bits(other)
-        return self * other
+        """ Bitwise and &. Cost: 1 constraint per bit """
+        if is_base_value(other):
+            other = other * LinComb.ONE
+
+        if (self.value | 1 == 1) and (other.value | 1 == 1):
+            return self * other
+
+        self_bits = self.to_bits()
+        other_bits = other.to_bits()
+        res = [self_bits[i] * other_bits[i] for i in range(bitlength)]
+
+        return LinComb.from_bits(res)
 
```

Comment:
Unfortunately, this does not work. The steps of the computation you perform, should be independent if the values you are computing on. Otherwise, the computation leaks information about the values, and you need different key material every time you are doing the computation. That is not the case here: this code does a 1-bit and for 1-bit values and a k-bit otherwise. Thus, it leaks whether the value it is called on, is a single bit. (As a rule, when writing oblivious code you should never need to access .value.)

Since applying binary operators on 1-bit values is quite common and in this case you don't what the overhead of doing a k-bit and, I think it is probably better to leave __and__, __or__, etc, to work on single-bit values. But having separate member functions for k-bit binary operators (e.g., value.and_n(other)) sounds like a good idea.

Does that make sense?

Reply:
I see. Thanks for the info! I'll rethink how to implement this.